Number_types/include/CGAL/FPU.h

```diff
@@ -107,7 +107,7 @@ extern "C" {
 #if  defined(__SSE2__) \
   || (defined(_M_IX86_FP) && _M_IX86_FP >= 2) \
   || defined(_M_X64)
-#  include <emmintrin.h>
+#  include <immintrin.h>
```

Comment:
I don't think so, `immintrin.h`, like `x86intrin.h`, is supposed to be the general "include all the intrinsic headers you can".

Reply:
In gcc, `<immintrin.h>` [appeared in 2008](https://github.com/gcc-mirror/gcc/commits/master?after=c914ddb0d259fe209117c28034d53d81f97dd702+34&path%5B%5D=gcc&path%5B%5D=config&path%5B%5D=i386&path%5B%5D=immintrin.h). In clang, it was [in 2010](https://github.com/llvm-mirror/clang/commit/033b7b327e8cdc8cd4bafce1625ea88bc9e3f6c8).

So far we support gcc and clang versions without the header.

Reply:
Then postpone the patch until CGAL moves into this decade? I am not interested in supporting those compilers.

Reply:
You lack pragmatism, here. With of few lines of preprocessor conditionals, one can solve the issue:
```c++
#  include <emmintrin.h>
#  if defined __SSE3__
#    include <pmmintrin.h>
#  endif
#  if defined __SSE4_1__
#    include <smmintrin.h>
#  endif
#  if defined __AVX__
#    include <immintrin.h>
#  endif
```

Just copy-paste those lines, or allow me to edit your pull-request.

Otherwise the pull-request will break something in the testsuite because, as you know, `__SSE2__` is by default on x86_64, and for any old compiler tested on x86_64, there will be a preprocessor error.

Reply:
Edit if you want, but does your code actually work? I think clang had support for __AVX__ before immintrin.h. Do all the visual studio you support have immintrin.h?

Reply:
Let's keep that simple and do not care about the support of __AVX__ before `<immintrin.h>`. My patch is anyway an improvement over the current state of the pull-request.

As for MSVC 2013, it already supported [`_mm_addsub_pd` and `_mm_blendv_pd`][2], but not `_mm_sqrt_round_sd`. [MSVC 2017 does not have it either][1]. Anyway, MSVC does not define `__AVX512F__` but `__AVX2__`

[1]: https://docs.microsoft.com/en-us/cpp/intrinsics/x64-amd64-intrinsics-list?view=vs-2017
[2]: https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/hh977023(v=vs.120)

Reply:
I mean: MSVC 2017 does not even have the compiler flag to opt for AVX512.

Reply:
I expect that, when we move to C++14, we will drop support for enough compilers that we can revert this and always include immintrin.h :wink:

Reply:
Yes, that header was added in gcc-4.0, and clang 2.8, and was already in VC 2012. Compilers supporting C++11/14 are more recent. But that will be a petty advantage of switching to C++14, compared to... all of C++14!